### ISSUE TITLE
deadlock at composer.trainer.trainer: Spinning the dataloaders

### ISSUE THREAD

**Issue by guoyejun** [mosaicml/llm-foundry#436]
## Environment
```
Collecting system information...
---------------------------------
System Environment Report
Created: 2023-07-07 01:31:32 PDT
---------------------------------

PyTorch information
-------------------
PyTorch version: 2.0.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.26.3
Libc version: glibc-2.31

Python version: 3.9.7 (default, Sep 16 2021, 13:09:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-149-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB
GPU 4: NVIDIA A100-SXM4-40GB
GPU 5: NVIDIA A100-SXM4-40GB
GPU 6: NVIDIA A100-SXM4-40GB
GPU 7: NVIDIA A100-SXM4-40GB

Nvidia driver version: 525.105.17
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   43 bits physical, 48 bits virtual
CPU(s):                          256
On-line CPU(s) list:             0-255
Thread(s) per core:              2
Core(s) per socket:              64
Socket(s):                       2
NUMA node(s):                    8
Vendor ID:                       AuthenticAMD
CPU family:                      23
Model:                           49
Model name:                      AMD EPYC 7742 64-Core Processor
Stepping:                        0
Frequency boost:                 enabled
CPU MHz:                         1706.139
CPU max MHz:                     2250.0000
CPU min MHz:                     1500.0000
BogoMIPS:                        4491.70
Virtualization:                  AMD-V
L1d cache:                       4 MiB
L1i cache:                       4 MiB
L2 cache:                        64 MiB
L3 cache:                        512 MiB
NUMA node0 CPU(s):               0-15,128-143
NUMA node1 CPU(s):               16-31,144-159
NUMA node2 CPU(s):               32-47,160-175
NUMA node3 CPU(s):               48-63,176-191
NUMA node4 CPU(s):               64-79,192-207
NUMA node5 CPU(s):               80-95,208-223
NUMA node6 CPU(s):               96-111,224-239
NUMA node7 CPU(s):               112-127,240-255
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate sme ssbd mba sev ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif umip rdpid overflow_recov succor smca

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] pytorch-ranger==0.1.1
[pip3] torch==2.0.1
[pip3] torch-optimizer==0.3.0
[pip3] torchdata==0.6.1
[pip3] torchmetrics==0.11.4
[pip3] torchtext==0.15.2
[pip3] torchvision==0.15.2
[conda] numpy                     1.24.4                   pypi_0    pypi
[conda] pytorch-ranger            0.1.1                    pypi_0    pypi
[conda] torch                     2.0.1                    pypi_0    pypi
[conda] torch-optimizer           0.3.0                    pypi_0    pypi
[conda] torchdata                 0.6.1                    pypi_0    pypi
[conda] torchmetrics              0.11.4                   pypi_0    pypi
[conda] torchtext                 0.15.2                   pypi_0    pypi
[conda] torchvision               0.15.2                   pypi_0    pypi


Composer information
--------------------
Composer version: 0.15.1
Composer commit hash: None
Host processor model name: AMD EPYC 7742 64-Core Processor
Host processor core count: 128
Number of nodes: 1
Accelerator model name: NVIDIA A100-SXM4-40GB
Accelerators per node: 1
CUDA Device Count: 8

```

## To reproduce
get the dataset ,and then use the exactly same command at https://github.com/mosaicml/llm-foundry/tree/main#quickstart
```
composer train/train.py   train/yamls/pretrain/mpt-125m.yaml   data_local=my-copy-c4   train_loader.dataset.split=train_small   eval_loader.dataset.split=val_small   max_duration=10ba   eval_interval=0   save_folder=mpt-125m
```

and we'll see the output as below:
```
...
Starting training...
2023-07-07 01:36:02,640: rank0[1074804][MainThread]: INFO: composer.trainer.trainer: Using precision Precision.AMP_BF16
******************************
Config:
enabled_algorithms/GradientClipping: true
node_name: unknown because NODENAME environment variable not set
num_gpus_per_node: 8
num_nodes: 1
rank_zero_seed: 17

******************************
2023-07-07 01:36:02,649: rank0[1074804][MainThread]: DEBUG: composer.trainer.trainer: Spinning the dataloaders
```

I waited for a long time and no progress, and it can be stopped with Ctrl+C.

At this time, I checked the dataset and found that most files have been deleted.
```
$ tree my-copy-c4/
my-copy-c4/
├── train_small
│   └── index.json
└── val_small
    └── index.json
```

I've tried to add 'keep_zip: true' in mpt-125m.yaml, but no use.

I also tried to change the mode size from 12 layers to 1 layers, and tried on two GPUs with below commands:
```
#1. re-get the data set
tree my-copy-c4/
my-copy-c4/
├── train_small
│   ├── index.json
│   ├── shard.00000.mds.zstd
│   ├── shard.00001.mds.zstd
│   ├── shard.00002.mds.zstd
│   ├── shard.00003.mds.zstd
│   ├── shard.00004.mds.zstd
│   ├── shard.00005.mds.zstd
│   ├── shard.00006.mds.zstd
│   ├── shard.00007.mds.zstd
│   ├── shard.00008.mds.zstd
│   ├── shard.00009.mds.zstd
│   ├── shard.00010.mds.zstd
│   ├── shard.00011.mds.zstd
│   ├── shard.00012.mds.zstd
│   ├── shard.00013.mds.zstd
│   ├── shard.00014.mds.zstd
│   ├── shard.00015.mds.zstd
│   ├── shard.00016.mds.zstd
│   ├── shard.00017.mds.zstd
│   ├── shard.00018.mds.zstd
│   ├── shard.00019.mds.zstd
│   ├── shard.00020.mds.zstd
│   ├── shard.00021.mds.zstd
│   ├── shard.00022.mds.zstd
│   ├── shard.00023.mds.zstd
│   └── shard.00024.mds.zstd
└── val_small
    ├── index.json
    ├── shard.00000.mds.zstd
    ├── shard.00001.mds.zstd
    └── shard.00002.mds.zstd

2 directories, 30 files
```
```
#2 the the command with 2 GPUs and with verbose. 
composer -n 2 -v train/train.py   train/yamls/pretrain/mpt-125m.yaml   data_local=my-copy-c4   train_loader.dataset.split=train_small   eval_loader.dataset.split=val_small   max_duration=10ba   eval_interval=0   save_folder=mpt-125m
```

and this time, I saw the crash as below.
```
Starting training...
2023-07-07 01:42:36,789: rank0[1085371][MainThread]: INFO: composer.trainer.trainer: Using precision Precision.AMP_BF16
******************************
Config:
enabled_algorithms/GradientClipping: true
node_name: unknown because NODENAME environment variable not set
num_gpus_per_node: 2
num_nodes: 1
rank_zero_seed: 17

******************************
2023-07-07 01:42:36,800: rank0[1085371][MainThread]: DEBUG: composer.trainer.trainer: Spinning the dataloaders
2023-07-07 01:42:37,380: rank0[1086841][ThreadPoolExecutor-0_0]: ERROR: concurrent.futures: exception calling callback for <Future at 0x7f70f0254160 state=finished raised RuntimeError>
Traceback (most recent call last):
  File "/raid/miniconda3/envs/yjguo_mpt/lib/python3.9/site-packages/streaming/base/stream.py", line 275, in _download_file
    download_file(remote, local, self.download_timeout)
  File "/raid/miniconda3/envs/yjguo_mpt/lib/python3.9/site-packages/streaming/base/storage/download.py", line 285, in download_file
    raise ValueError(
ValueError: In the absence of local dataset, path to remote dataset must be provided

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/raid/miniconda3/envs/yjguo_mpt/lib/python3.9/concurrent/futures/_base.py", line 329, in _invoke_callbacks
    callback(self)
  File "/raid/miniconda3/envs/yjguo_mpt/lib/python3.9/site-packages/streaming/base/dataset.py", line 1086, in on_exception
    raise exception
  File "/raid/miniconda3/envs/yjguo_mpt/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/raid/miniconda3/envs/yjguo_mpt/lib/python3.9/site-packages/streaming/base/dataset.py", line 1132, in _download_thread
    self.download_shard(shard_id, False)
  File "/raid/miniconda3/envs/yjguo_mpt/lib/python3.9/site-packages/streaming/base/dataset.py", line 985, in download_shard
    stream.download_shard(shard)
  File "/raid/miniconda3/envs/yjguo_mpt/lib/python3.9/site-packages/streaming/base/stream.py", line 366, in download_shard
    self._download_shard_part(raw_info, zip_info, shard.compression)
  File "/raid/miniconda3/envs/yjguo_mpt/lib/python3.9/site-packages/streaming/base/stream.py", line 345, in _download_shard_part
    self._download_file(zip_info.basename)
  File "/raid/miniconda3/envs/yjguo_mpt/lib/python3.9/site-packages/streaming/base/stream.py", line 284, in _download_file
    raise RuntimeError(
RuntimeError: Failed to download None -> my-copy-c4/val_small/shard.00002.mds.zstd. Tried 3 times, got errors:
[ValueError('In the absence of local dataset, path to remote dataset must be provided'), ValueError('In the absence of local dataset, path to remote dataset must be provided'), ValueError('In the absence of local dataset, path to remote dataset must be provided')]

```

**Comment by guoyejun** [comment]
btw, I have network issue and could not download the data again,  and so have to use the local files but they are deleted in the script running.

**Comment by guoyejun** [comment]
I now found that the issue is relative to the compression of the dataset, the issue is gone if I remove "--compression zstd" at https://github.com/mosaicml/llm-foundry/tree/main/scripts/data_prep#huggingface-data

**Comment by guoyejun** [comment]
with "--compression zstd" removed, I successfully debug the script for some times, but suddenly the issue appeared again (see below). I never touch the data files manually during the period.  I have tried to uninstall/install mosaicml, mosaicml-streaming, and llm-foundry, but no use, still the same issue.  Once I see the issue, I always see the issue.

```
ValueError: Reused local directory: ['/raid/yjguo/mpt/llm-foundry/scripts/my-copy-c4/train_small'] vs
['/raid/yjguo/mpt/llm-foundry/scripts/my-copy-c4/train_small']. Provide a different one. If using a unique local directory, try
deleting the local directory and call `streaming.base.util.clean_stale_shared_memory()` only once in your script to clean up the
stale shared memory before instantiation of `StreamingDataset`.
```

I've tried call streaming.base.util.clean_stale_shared_memory() in train.py, but no use.

**Comment by knighton** [comment]
I am so sorry to hear that.

Streaming uses several data structures that live in in shared memory in order to coordinate worker processes (eg, jointly managing the shard cache(s)).

There unfortunately are ways for processes to die or go into a zombie state that leaves shared memory objects behind. We have been improving this situation over time, but our work here is clearly not over. 

Because the pain of debugging race conditions due to shared resources being modified (especially unknowingly) by multiple runs at the same time is even worse, we added strict local dir collision checks.

I see you are running on Ubuntu. You should be able to list `/dev/shm/` and find whatever streaming bookkeeping that should not be there (it will end with `_locals`) and can be removed. Feel free to reply with the listing and we can write you the script.

If you can reboot the system, that will clear away all shared memory too.

**Comment by guoyejun** [comment]
thanks, the issue is fixed with "rm -rf /dev/shm/*" and prepare the data files again.

**Comment by guoyejun** [comment]
thanks, the issue is gone after machine reboot.

**Comment by guoyejun** [comment]
btw, the issue is still there if I download the dataset with "--compression zstd"

**Comment by mayankjobanputra** [comment]
> btw, the issue is still there if I download the dataset with "--compression zstd"

I am facing the same issue. Any clue on how to fix this?

**Comment by karan6181** [comment]
@guoyejun What would happen if you run this `composer train/train.py train/yamls/pretrain/mpt-125m.yaml   data_remote=my-copy-c4 data_local=/tmp/streaming/dataset  train_loader.dataset.split=train_small   eval_loader.dataset.split=val_small   max_duration=10ba   eval_interval=0   save_folder=mpt-125m`. Atleast, you should not see the dataset getting deleted. The changes here are, provide `remote` location as a current local directory location and a `local` as a new directory. Streaming dataset would copy the data from remote to local.  

@knighton is working on a fix as part of PR https://github.com/mosaicml/streaming/pull/330

**Comment by dakinggg** [comment]
I believe this has been fixed by the linked pull request. Please open a new issue if you are still encountering problems.

**Comment by LingxiaoShawn** [comment]
This problem is still there, and it can be solved with clear the cache mentioned above.
